homeassistant/components/synology_dsm/button.py

```diff
@@ -0,0 +1,96 @@
+"""Support for Synology DSM buttons."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+import logging
+from typing import Any, Final
+
+from homeassistant.components.button import (
+    ButtonDeviceClass,
+    ButtonEntity,
+    ButtonEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import DeviceInfo, EntityCategory
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import SynoApi
+from .const import DOMAIN, SYNO_API
+
+LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class SynologyDSMbuttonDescriptionMixin:
+    """Mixin to describe a Synology DSM button entity."""
+
+    press_action: Callable[[SynoApi], Any]
+
+
+@dataclass
+class SynologyDSMbuttonDescription(
+    ButtonEntityDescription, SynologyDSMbuttonDescriptionMixin
+):
+    """Class to describe a Synology DSM button entity."""
+
+
+BUTTONS: Final = [
+    SynologyDSMbuttonDescription(
+        key="reboot",
+        name="Reboot",
+        device_class=ButtonDeviceClass.RESTART,
+        entity_category=EntityCategory.CONFIG,
+        press_action=lambda syno_api: syno_api.async_reboot(),
+    ),
+    SynologyDSMbuttonDescription(
+        key="shutdown",
+        name="Shutdown",
+        icon="mdi:power",
+        entity_category=EntityCategory.CONFIG,
+        press_action=lambda syno_api: syno_api.async_shutdown(),
+    ),
+]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set buttons for device."""
+    data = hass.data[DOMAIN][entry.unique_id]
+    syno_api: SynoApi = data[SYNO_API]
+
+    async_add_entities([SynologyDSMButton(syno_api, button) for button in BUTTONS])
+
+
+class SynologyDSMButton(ButtonEntity):
+    """Defines a Synology DSM button."""
+
+    entity_description: SynologyDSMbuttonDescription
+
+    def __init__(
+        self,
+        api: SynoApi,
+        description: SynologyDSMbuttonDescription,
+    ) -> None:
+        """Initialize the Synology DSM binary_sensor entity."""
+        self.entity_description = description
+        self.syno_api = api
+
+        self._attr_name = f"{api.network.hostname} {description.name}"
+        self._attr_unique_id = f"{api.information.serial}_{description.key}"
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, api.information.serial)}
+        )
```

Comment:
It would be good to make a base class that sets the device info that is a subset of `SynologyDSMBaseEntity` so we don't have the code duplicated here.

Reply:
I thought about inheriting from `SynologyDSMBaseEntity` but there is to much in it, which is not needed by a simple button entity. SO I decided to add this minimum of device_info to get the buttons connected to the device created by the integration.

Reply:
It would be cleaner to split up `SynologyDSMBaseEntity` to avoid the code duplication.

This PR is already large enough though so it can wait for the future.

Reply:
I will do this in a follow-up PR soon 👍